iree/compiler/Dialect/Stream/Transforms/ScheduleAllocation.cpp

```diff
@@ -0,0 +1,1440 @@
+      // Primitive operand.
+      newOperands.push_back(operand);
+      continue;
+    }
+
+    // Read-only or read-write. Write-only are untied results below.
+    unsigned operandIdx =
+        asyncOp.getTiedOperandsIndexAndLength().first + it.index();
+    auto accessBits = IREE::Stream::ResourceAccessBitfield::Read;
+    if (asyncOp.isOperandTied(operandIdx)) {
+      accessBits = accessBits | IREE::Stream::ResourceAccessBitfield::Write;
+    }
+
+    auto resourceRange = scope.lookupResourceRange(operand);
+    auto resourceOffset = resourceRange.offset;
+    auto resourceLength = asyncOp.getOperandSize(operandIdx);
+    auto resourceAccess = IREE::Stream::ResourceAccessBitfieldAttr::get(
+        builder.getContext(), accessBits);
+    newResources.push_back(resourceRange.resource);
+    newResourceSizes.push_back(resourceRange.resourceSize);
+    newResourceOffsets.push_back(resourceOffset);
+    newResourceLengths.push_back(resourceLength);
+    newResourceAccesses.push_back(resourceAccess);
+  }
+
+  for (auto result : asyncOp.results()) {
+    auto tiedOperand = asyncOp.getTiedResultOperand(result);
+    if (tiedOperand) {
+      // All tied results are handled above as read-write.
+      continue;
+    }
+    auto resourceRange = scope.lookupResourceRange(result);
+    auto resourceOffset = resourceRange.offset;
+    auto resourceLength = asyncOp.getResultSize(result.getResultNumber());
+    auto resourceAccess = IREE::Stream::ResourceAccessBitfieldAttr::get(
+        builder.getContext(), IREE::Stream::ResourceAccessBitfield::Write);
+    newResources.push_back(resourceRange.resource);
+    newResourceSizes.push_back(resourceRange.resourceSize);
+    newResourceOffsets.push_back(resourceOffset);
+    newResourceLengths.push_back(resourceLength);
+    newResourceAccesses.push_back(resourceAccess);
+  }
+
+  builder.create<IREE::Stream::CmdDispatchOp>(
+      asyncOp.getLoc(), asyncOp.workgroup_count(), asyncOp.entry_point(),
+      newOperands, newResources, newResourceSizes, newResourceOffsets,
+      newResourceLengths, builder.getArrayAttr(newResourceAccesses));
+  asyncOp.erase();
+  return success();
+}
+
+static LogicalResult applyAsyncAllocations(Region &region,
+                                           AllocationScope &scope);
+
+static LogicalResult applyAsyncConcurrentOp(
+    IREE::Stream::AsyncConcurrentOp asyncOp, AllocationScope &scope,
+    OpBuilder builder) {
+  // Remove operands from the yield now that we aren't returning anything.
+  // Must do this before we recurse so that the ops we are transforming have no
+  // uses.
+  auto yieldOp =
+      cast<IREE::Stream::YieldOp>(asyncOp.body().front().getTerminator());
+  yieldOp->eraseOperands(0, yieldOp.getNumOperands());
+
+  // Recurse into the wave ops to process the async ops within.
+  // They'll be mapped to resources in the parent scope - we must fix this up!
+  if (failed(applyAsyncAllocations(asyncOp.body(), scope))) {
+    return failure();
+  }
```

Comment:
Is this "we must fix this up!" referring to something not yet implemented (add TODO)?

Reply:
ah, it's just noting that the resources are mapped in the parent scope and we must fix it up - will clarify